pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,483 @@
+
+        private static final Recycler<AsyncAddArgs> ASYNC_ADD_ARGS_RECYCLER = new Recycler<>() {
+            @Override
+            protected AsyncAddArgs newObject(Handle<AsyncAddArgs> handle) {
+                return new AsyncAddArgs(handle);
+            }
+        };
+
+        public static AsyncAddArgs newInstance(AddDataCallback callback, Object ctx, long addedTime){
+            AsyncAddArgs asyncAddArgs = ASYNC_ADD_ARGS_RECYCLER.get();
+            asyncAddArgs.callback = callback;
+            asyncAddArgs.ctx = ctx;
+            asyncAddArgs.addedTime = addedTime;
+            return asyncAddArgs;
+        }
+
+        private AsyncAddArgs(Recycler.Handle<AsyncAddArgs> handle){
+            this.handle = handle;
+        }
+
+        private final Recycler.Handle<AsyncAddArgs> handle;
+
+        /** Argument for {@link #asyncAddData(Object, AddDataCallback, Object)}. **/
+        @Getter
+        private AddDataCallback callback;
+
+        /** Argument for {@link #asyncAddData(Object, AddDataCallback, Object)}. **/
+        @Getter
+        private Object ctx;
+
+        /** Time of executed {@link #asyncAddData(Object, AddDataCallback, Object)}. **/
+        @Getter
+        private long addedTime;
+
+    }
+
+    /***
+     * The context for {@link ManagedLedger#asyncAddEntry(byte[], AsyncCallbacks.AddEntryCallback, Object)}, Holds the
+     * data array written in batches and callback parameters that need to be executed after batched write is complete.
+     */
+    public static class FlushContext{
+
+        private static final Recycler<FlushContext> FLUSH_CONTEXT_RECYCLER = new Recycler<FlushContext>() {
+            @Override
+            protected FlushContext newObject(Handle<FlushContext> handle) {
+                return new FlushContext(handle);
+            }
+        };
+
+        private final Recycler.Handle<FlushContext> handle;
+
+        /** Callback parameters for current batch. **/
+        private final ArrayList<AsyncAddArgs> asyncAddArgsList;
+
+        public FlushContext(Recycler.Handle<FlushContext> handle){
+            this.handle = handle;
+            this.asyncAddArgsList = new ArrayList<>(8);
+        }
+
+        public static FlushContext newInstance(){
+            return FLUSH_CONTEXT_RECYCLER.get();
+        }
+
+        public static FlushContext newInstance(int asyncAddArgsListInitialCapacity){
+            FlushContext flushContext = FLUSH_CONTEXT_RECYCLER.get();
+            flushContext.asyncAddArgsList.ensureCapacity(asyncAddArgsListInitialCapacity);
+            return flushContext;
+        }
+
+        public void recycle(){
+            this.asyncAddArgsList.clear();
+            this.handle.recycle(this);
+        }
+    }
+
+    /***
+     * The difference with {@link PositionImpl} is that there are two more parameters:
+     * {@link #batchSize}, {@link #batchIndex}.
+     */
+    public static class TxnBatchedPositionImpl extends PositionImpl {
```

Comment:
If this one intended to be exposed to the outside, please move out from this class, just create a separate public class.

Reply:
Already Fixed.